File: packages/dart/lib/Seguranca/Ataques/AtaqueDiferencialKeystream.dart
import 'dart:typed_data';

import 'package:criptografia/Seguranca/AlvoCriptografico.dart';
import 'package:criptografia/Seguranca/Ataque.dart';
import 'package:criptografia/Seguranca/ResultadoAtaque.dart';
import 'package:criptografia/Seguranca/Util.dart';

/// Criptanálise diferencial no keystream: com uma diferença FIXA (1 bit) no
/// IV, coleta as diferenças de saída entre ks(iv) e ks(iv ^ delta). Numa cifra
/// boa, essas diferenças são uniformes e únicas. Sinaliza:
///  - bits de saída que NUNCA mudam (ou mudam sempre) para essa diferença;
///  - diferenças de saída que se REPETEM (diferencial de alta probabilidade),
///    que é o insumo básico de um ataque diferencial.
class AtaqueDiferencialKeystream extends Ataque {
  final int _amostras;
  final int _tamanhoBloco;

  AtaqueDiferencialKeystream({this._amostras = 2000, this._tamanhoBloco = 32});

  @override
  String nome() => 'Diferencial do keystream (delta fixo no IV)';

  @override
  ResultadoAtaque executar(AlvoCriptografico alvo) {
    final chave = randomBytes(alvo.chaveDeTeste().length);
    final tamanhoIv = alvo.tamanhoIv();

    final delta = Uint8List(tamanhoIv);
    delta[randomInt(0, tamanhoIv - 1)] = 1 << randomInt(0, 7);

    final totalBits = _tamanhoBloco * 8;
    final sempreZero = List<bool>.filled(totalBits, true);
    final sempreUm = List<bool>.filled(totalBits, true);
    final deltas = <String>{};

    for (var i = 0; i < _amostras; i++) {
      final iv = randomBytes(tamanhoIv);
      final iv2 = _xorBytes(iv, delta);

      final ks1 = alvo.gerarKeystreamBruto(
        chave,
        iv,
        'enc',
        _tamanhoBloco,
      );
      final ks2 = alvo.gerarKeystreamBruto(
        chave,
        iv2,
        'enc',
        _tamanhoBloco,
      );
      final d = _xorBytes(ks1, ks2);
      deltas.add(bytesToHex(d));

      for (var p = 0; p < d.length; p++) {
        final byte = d[p];
        for (var b = 0; b < 8; b++) {
          final idx = p * 8 + b;
          if (((byte >> b) & 1) == 1) {
            sempreZero[idx] = false;
          } else {
            sempreUm[idx] = false;
          }
        }
      }
    }

    var bitsFixos = 0;
    for (var idx = 0; idx < totalBits; idx++) {
      if (sempreZero[idx] || sempreUm[idx]) {
        bitsFixos++;
      }
    }
    final colisoes = _amostras - deltas.length;

    final vulneravel = bitsFixos > 0 || colisoes > 0;

    return ResultadoAtaque(
      nome(),
      vulneravel,
      vulneravel ? Severidade.alta : Severidade.info,
      'delta de 1 bit no IV: $bitsFixos bit(s) de saída fixo(s), '
          '$colisoes diferencial(is) repetido(s) em $_amostras amostras',
      {'bits_fixos': bitsFixos, 'colisoes': colisoes},
    );
  }

  Uint8List _xorBytes(Uint8List a, Uint8List b) {
    final out = Uint8List(a.length);
    for (var i = 0; i < a.length; i++) {
      out[i] = a[i] ^ b[i];
    }
    return out;
  }
}

File: packages/rust/src/seguranca/ataques/ataque_diferencial_keystream.rs
//! Criptanálise diferencial no keystream: com uma diferença FIXA (1 bit) no
//! IV, coleta as diferenças de saída entre ks(iv) e ks(iv ^ delta). Numa cifra
//! boa, essas diferenças são uniformes e únicas. Sinaliza:
//!  - bits de saída que NUNCA mudam (ou mudam sempre) para essa diferença;
//!  - diferenças de saída que se REPETEM (diferencial de alta probabilidade),
//!    que é o insumo básico de um ataque diferencial.

use std::collections::HashSet;

use crate::seguranca::alvo_criptografico::AlvoCriptografico;
use crate::seguranca::ataque::{Ataque, ErroAtaque};
use crate::seguranca::resultado_ataque::{ResultadoAtaque, Severidade};
use crate::seguranca::util::{hex_encode, random_bytes, random_int};

pub struct AtaqueDiferencialKeystream {
    pub amostras: usize,
    pub tamanho_bloco: usize,
}

impl AtaqueDiferencialKeystream {
    pub fn new() -> Self {
        AtaqueDiferencialKeystream {
            amostras: 2000,
            tamanho_bloco: 32,
        }
    }
}

impl Default for AtaqueDiferencialKeystream {
    fn default() -> Self {
        Self::new()
    }
}

impl Ataque for AtaqueDiferencialKeystream {
    fn nome(&self) -> String {
        "Diferencial do keystream (delta fixo no IV)".to_string()
    }

    fn executar(&self, alvo: &dyn AlvoCriptografico) -> Result<ResultadoAtaque, ErroAtaque> {
        let chave = random_bytes(alvo.chave_de_teste().len());
        let tamanho_iv = alvo.tamanho_iv();

        let mut delta = vec![0u8; tamanho_iv];
        delta[random_int(0, tamanho_iv as i64 - 1) as usize] =
            1u8 << (random_int(0, 7) as u32);

        let total_bits = self.tamanho_bloco * 8;
        let mut sempre_zero = vec![true; total_bits];
        let mut sempre_um = vec![true; total_bits];
        let mut deltas: HashSet<String> = HashSet::new();

        for _ in 0..self.amostras {
            let iv = random_bytes(tamanho_iv);
            let iv2: Vec<u8> = iv.iter().zip(delta.iter()).map(|(a, b)| a ^ b).collect();

            let ks1 =
                alvo.gerar_keystream_bruto(&chave, &iv, "enc", self.tamanho_bloco);
            let ks2 =
                alvo.gerar_keystream_bruto(&chave, &iv2, "enc", self.tamanho_bloco);
            let d: Vec<u8> = ks1.iter().zip(ks2.iter()).map(|(a, b)| a ^ b).collect();
            deltas.insert(hex_encode(&d));

            for p in 0..d.len() {
                let byte = d[p];
                for b in 0..8 {
                    let idx = p * 8 + b;
                    if ((byte >> b) & 1) == 1 {
                        sempre_zero[idx] = false;
                    } else {
                        sempre_um[idx] = false;
                    }
                }
            }
        }

        let mut bits_fixos = 0usize;
        for idx in 0..total_bits {
            if sempre_zero[idx] || sempre_um[idx] {
                bits_fixos += 1;
            }
        }
        let colisoes = self.amostras - deltas.len();

        let vulneravel = bits_fixos > 0 || colisoes > 0;

        Ok(ResultadoAtaque::new(
            &self.nome(),
            vulneravel,
            if vulneravel {
                Severidade::Alta
            } else {
                Severidade::Info
            },
            &format!(
                "delta de 1 bit no IV: {} bit(s) de saída fixo(s), {} diferencial(is) repetido(s) em {} amostras",
                bits_fixos, colisoes, self.amostras
            ),
        ))
    }
}
